storage/memorybysubnet/peer_store.go

```diff
@@ -0,0 +1,595 @@
+	if cfg.ShardCount > 0 {
+		shardCount = cfg.ShardCount
+	}
+
+	if cfg.GarbageCollectionInterval <= 0 {
+		return nil, ErrInvalidGCInterval
+	}
+
+	ps := &peerStore{
+		shards:   make([]*peerShard, shardCount*2),
+		closed:   make(chan struct{}),
+		ipv4Mask: cfg.PreferredIPv4SubnetMaskBits,
+		ipv6Mask: cfg.PreferredIPv6SubnetMaskBits,
+	}
+
+	for i := 0; i < shardCount*2; i++ {
+		ps.shards[i] = &peerShard{swarms: make(map[bittorrent.InfoHash]swarm)}
+	}
+
+	go func() {
+		for {
+			select {
+			case <-ps.closed:
+				return
+			case <-time.After(cfg.GarbageCollectionInterval):
+				before := time.Now().Add(-cfg.PeerLifetime)
+				log.Debugln("memory: purging peers with no announces since", before)
+				ps.collectGarbage(before)
+			}
+		}
+	}()
+
+	return ps, nil
+}
+
+type serializedPeer string
+
+type peerShard struct {
+	swarms map[bittorrent.InfoHash]swarm
+	sync.RWMutex
+}
+
+type swarm struct {
+	// map serialized peer to mtime
+	seeders  map[string]map[serializedPeer]int64
+	leechers map[string]map[serializedPeer]int64
+}
+
+func (s swarm) lenSeeders() (i int) {
+	for _, subnet := range s.seeders {
+		for range subnet {
+			i++
+		}
+	}
+	return
+}
+
+func (s swarm) lenLeechers() (i int) {
+	for _, subnet := range s.leechers {
+		for range subnet {
+			i++
+		}
+	}
+	return
+}
+
+type peerStore struct {
+	shards   []*peerShard
+	closed   chan struct{}
+	ipv4Mask int
+	ipv6Mask int
+}
+
+var _ storage.PeerStore = &peerStore{}
+
+func (s *peerStore) shardIndex(infoHash bittorrent.InfoHash, af bittorrent.AddressFamily) uint32 {
+	// There are twice the amount of shards specified by the user, the first
+	// half is dedicated to IPv4 swarms and the second half is dedicated to
+	// IPv6 swarms.
+	idx := binary.BigEndian.Uint32(infoHash[:4]) % (uint32(len(s.shards)) / 2)
+	if af == bittorrent.IPv6 {
+		idx += uint32(len(s.shards) / 2)
+	}
+	return idx
+}
+
+func newPeerKey(p bittorrent.Peer) serializedPeer {
+	b := make([]byte, 20+2+len(p.IP.IP))
+	copy(b[:20], p.ID[:])
+	binary.BigEndian.PutUint16(b[20:22], p.Port)
+	copy(b[22:], p.IP.IP)
+
+	return serializedPeer(b)
+}
+
+func decodePeerKey(pk serializedPeer) bittorrent.Peer {
+	peer := bittorrent.Peer{
+		ID:   bittorrent.PeerIDFromString(string(pk[:20])),
+		Port: binary.BigEndian.Uint16([]byte(pk[20:22])),
+		IP:   bittorrent.IP{IP: net.IP(pk[22:])}}
+
+	if ip := peer.IP.To4(); ip != nil {
+		peer.IP.IP = ip
+		peer.IP.AddressFamily = bittorrent.IPv4
+	} else if len(peer.IP.IP) == net.IPv6len { // implies toReturn.IP.To4() == nil
+		peer.IP.AddressFamily = bittorrent.IPv6
+	} else {
+		panic("IP is neither v4 nor v6")
+	}
+
+	return peer
+}
+
+func (s *peerStore) mask(p bittorrent.Peer) string {
+	var maskedIP net.IP
+	switch p.IP.AddressFamily {
+	case bittorrent.IPv4:
+		maskedIP = p.IP.IP.Mask(net.CIDRMask(s.ipv4Mask, 32))
+	case bittorrent.IPv6:
+		maskedIP = p.IP.IP.Mask(net.CIDRMask(s.ipv6Mask, 128))
+	default:
+		panic("IP is neither v4 nor v6")
+	}
+	return maskedIP.String()
```

Comment:
Masks are slices of bytes, which means they're dynamically allocated and can't be used as keys in maps 😿